### PR TITLE
ramips: add support for SNR-CPE-ME2-SFP

### PR DESCRIPTION
**Pull request by alealexpro100**
SNR-CPE-ME2-SFP is a wireless WiFi 5 router manufactured by SNR/NAG company.

```Specification:
- SoC: MediaTek MT7621A
- CPU: 880MHz
- Flash: 16 MB (GD25Q127CSIG)
- RAM:  256 MB
- WLAN: 2.4 GHz, 5 GHz (MediaTek MT7615DN)
- Ethernet: 4x 10/100/1000 Mbps
- SFP cage (using RTL8211FS-CG)
- USB 3.0 port
- Power: 12 VDC, 2 A
```

- [x] Update using sysupgrade
- [x] GPIO LED and BUTTON
- [x] Wi-Fi
- [x] SFP
- [x] Ethernet
- [x] USB
- [x] Stability
- [x] Correct MAC
- [x] U-boot env

Builds link: https://disk.yandex.ru/d/2Q_AgZliMr3e-Q

### CODE REVIEW

File: target/linux/ramips/image/mt7621.mk [hunk 1]
```diff
@@ -1937,6 +1937,18 @@ define Device/snr_snr-cpe-me2-lite
...
 TARGET_DEVICES += snr_snr-cpe-me2-lite
 
+define Device/snr_cpe-me2-sfp
+  $(Device/dsa-migration)
+  $(Device/uimage-lzma-loader)
+  IMAGE_SIZE := 16064k
+  DEVICE_VENDOR := SNR
+  DEVICE_MODEL := SNR-CPE-ME2-SFP
+  UIMAGE_NAME := $$(DEVICE_MODEL)
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt7615e kmod-mt7663-firmware-ap \
```
Comment (csharper2005) [inline target/linux/ramips/image/mt7621.mk:1947]: Why do you need kmod-mt7603, kmod-mt7615e, kmod-mt7663-firmware-ap? kmod-mt7615-firmware should be enough.

**Review by csharper2005 — CHANGES_REQUESTED** [review]

File: target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+
+		led {
+			led-active-low;
+		};
+	};
+};
+
+&pcie1 {
```
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:191]: You can drop this block for 7615dn.

**Comment by alealexpro100** [comment]
Removed note `reading sfp eeprom is NOT working` because it is driver issue https://github.com/openwrt/openwrt/issues/12045.

File: target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+			gpio-export,name="usb-power";
+			gpio-export,output=<1>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&state_default {
+	gpio {
+		groups = "jtag", "wdt";
```
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:69]: I would suggest checking it out. I think you need to add at least "uart" here. And maybe remove wdt.

File: target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+&pcie {
+	status = "okay";
+};
+
+&pcie0 {
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
```
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:184]: Are the wireless MACs correct? I would suggest that you add MAC addresses table to the commit message. Here is an example - https://github.com/openwrt/openwrt/commit/f490295bf213c5e181d16cbe53c163c1c9983e45

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:184]: Added MACs table

File: target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts (hunk 3 above)
Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:69]: UART is not used, but available to be soldered.

File: target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts (hunk 4 above)
Comment (csharper2005) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:184]: Are these real MACs from the stock firmware?
```
+----------+------------------+-------------------+
| use      | address          | example           |
+----------+------------------+-------------------+
| Device   | label            | 00:00:00:11:00:00 |
| Ethernet | + 1              | 00:00:00:11:00:01 |
| 2g       | + 2              | 00:00:00:11:00:02 |
| 5g       | + 3              | 00:00:00:11:00:03 |
+----------+------------------+-------------------+
```

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:184]: Yes.

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:184]: Iteration is from stock image. These MACs are written for example, they are not real.

File: target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts (hunk 3 above)
Comment (csharper2005) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts]: This is not about uart on the board. You should specify all pin groups used as gpio here.
https://www.t-firefly.com/download/FireWRT/hardware/MT7621.pdf -> 2.4.1 GPIO Pin share scheme

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts]: State_default is already written according to this datasheet and pins.

Comment (csharper2005) [reply target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts]: Thanks and sorry! I checked outdated version of your patch.

**Review by csharper2005 — APPROVED** [review]

File: target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+		};
+
+		led_vpn: led-1 {
+			label = "green:vpn";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		usb {
```
Comment (MaxS0niX) [inline target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:29]: https://www.kernel.org/doc/html/latest/leds/leds-class.html
I propose to generalize

File: target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+}; 
+
+&gdma {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 { // GD25Q127CSIG
```
Comment (MaxS0niX) [inline target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:81]: Here and in the commit message
https://gist.github.com/alealexpro100/8d63f271dba592c1121c1f98f7ccb529#file-snr-log2-L109

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:88]: The chip soldered on board is named GD25Q127CSIG. Output shows other.

File: target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+			label = "green:sys";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		vpn {
+			label = "green:vpn";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
```
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:32]: https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml
Look at `function-enumerator`. We have more than one instance of the same function.

File: target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+		led-upgrade = &led_sys;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_sys: sys {
+			label = "green:sys";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
```
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:25]: https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml
Look at function-enumerator. We have more than one instance of the same function.

**Review by csharper2005 — CHANGES_REQUESTED** [review]

File: target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		tx-disable-gpios = <&gpio 9 GPIO_ACTIVE_HIGH>;
+		maximum-power-milliwatt = <1000>;
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+		usb-power {
+			gpio-export,name="usb-power";
+			gpio-export,output=<1>;
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:68]: MT7621 USB power supply can be controlled by regulator. ref: https://github.com/openwrt/openwrt/commit/665c2154ef122d10dfffeca95538ebdff82653b5

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts]: Thanks. Added regulator for usb to dts.

Comment (csharper2005) [reply target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts]: > Thanks. Added regulator for usb to dts.

Is it works? Can you switch off and switch on the power on a real USB device?

Comment (alealexpro100) [reply target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts]: Yes, it works. Using `echo '1-1' | tee /sys/bus/usb/drivers/usb/unbind` and `echo '1-1' | tee /sys/bus/usb/drivers/usb/bind`.

Comment (csharper2005) [reply target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts]: Good news! Thanks!

**Review by csharper2005 — APPROVED** [review]

**Comment by hauke** [comment]
Do you plan to add support for reading and writing the u-boot environment soon?

You can backport the patches needed to make the SFP slot fully work. You can add them to target/linux/generic/backport-5.15/

**Comment by alealexpro100** [comment]
> Do you plan to add support for reading and writing the u-boot environment soon?

No. I did not have success to access U-boot (UART did not work). 

> You can backport the patches needed to make the SFP slot fully work. You can add them to target/linux/generic/backport-5.15/

Unfortunately, kernel (even [upstream](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/drivers/net/phy/realtek.c)) does not have support for it. I tried to use these patches ([1](https://github.com/DasSecurity-HatLab/openwrt/blob/ec05735c7181ec12fac948cb229ab0937f367020/target/linux/ramips/patches-5.4/9902-rtl8211fs.patch) and [2](https://github.com/x-wrt/x-wrt/blob/a8384ead787f4eee8752813dfaf646c9cc8ed3d6/target/linux/ramips/patches-5.10/992-net-phy-realtek-RTL8211FS-support-for-fiber-mode.patch)), but it did not work (reading eeprom).

**Pull request merged**